File: android/app/src/main/java/technology/tiny/app/chat/Attachments.kt
package technology.tiny.app.chat

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import android.provider.OpenableColumns
import android.util.Base64
import androidx.exifinterface.media.ExifInterface
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream

/** A picked document awaiting send (Converse document block). */
data class PendingDoc(val name: String, val format: String, val base64: String)

/**
 * Attachments → Converse content blocks (server lib/file-attachments.ts parity):
 *  - images: ≤4, downscaled 1568px JPEG q0.85 → {"image":{"format":"jpeg","source":{"bytes":…}}}
 *  - documents: ≤3MB raw bytes → {"document":{"name":safe,"format":…,"source":{"bytes":…}}}
 * Names are sanitized (Anthropic rejects odd chars): drop extension, non
 * [A-Za-z0-9 \-()[\]] → '_', cap 200.
 */
object Attachments {
    const val MAX = 4
    const val MAX_DOC_BYTES = 3_000_000

    /**
     * Human-readable per-document cap for the "documents must be under …"
     * reject message — DERIVED from MAX_DOC_BYTES, the same fix 13bd170 made for
     * the batch cap. Web renders it MiB (÷1024², one decimal): 3_000_000 B →
     * "2.9MB", NOT the decimal-MB "3MB" this message used to hardcode. That was
     * an internal self-contradiction: the same string reports the file's size in
     * MiB (÷1_048_576), so a 3_000_001 B file was rejected with "is 2.9MB —
     * documents must be under 3MB" — a stated limit READING HIGHER than the file
     * it just refused, and 0.1MB above web's "2.9MB". Compute it so the copy
     * stays honest and self-updates if the cap moves. (web lib/file-attachments.ts:130
     * `(MAX_DOCUMENT_BYTES/1024/1024).toFixed(1)`.)
     */
    val MAX_DOC_LABEL: String =
        String.format(java.util.Locale.US, "%.1fMB", MAX_DOC_BYTES / 1_048_576.0)

    /**
     * Total decoded-payload cap across all staged attachments (server
     * MAX_PAYLOAD_BYTES, lib/file-attachments.ts; iOS MAX_ATTACHMENTS_PAYLOAD_BYTES).
     * base64 inflates 4/3× and history + system context ride the same request,
     * so the body must stay under the worker's ~4.5MB cap. Per-item caps
     * (3MB/doc, downscaled images) don't stop FOUR heavy picks from summing past
     * it — this is the batch guard the mobile clients were missing (only web had
     * it), so an over-budget set was caught server-side as a send failure instead
     * of up front in the composer.
     */
    const val MAX_PAYLOAD_BYTES = 3_500_000

    /**
     * Human-readable payload cap for the "exceeds … total" composer message —
     * DERIVED from MAX_PAYLOAD_BYTES, not hardcoded. Web and iOS both format it
     * as MiB (÷1024², one decimal): 3_500_000 B → "3.3MB", NOT "3.5MB". The
     * message here used to hardcode "3.5MB" (a decimal-MB reading of the same
     * constant), so Android told the user a limit 0.2MB higher than the other
     * two clients and than the real byte cap — someone trimming to "just under
     * 3.5MB" would still be rejected. Compute it so the copy stays correct and
     * self-updates if the cap ever moves. (web Chat.tsx:715
     * `(MAX_PAYLOAD_BYTES/1024/1024).toFixed(1)`; iOS Views.swift:2625.)
     *
     * Pin Locale.US on the format like MAX_DOC_LABEL (:43) and the doc-size
     * messages (:215/:223): the reference is web's `.toFixed(1)`, which ALWAYS
     * emits a dot decimal. Without a locale, String.format uses the device
     * default, so a comma-decimal locale (tr/de/fr) rendered "3,3MB" here while
     * the same cap read "3.3MB" on web and in this file's other size labels —
     * one client silently disagreeing with the others on the very number a user
     * trims their attachment set toward.
     */
    val MAX_PAYLOAD_LABEL: String =
        String.format(java.util.Locale.US, "%.1fMB", MAX_PAYLOAD_BYTES / 1_048_576.0)

    /** Longest edge for the MODEL-bound image — the canonical cross-client value
     *  (web MAX_IMAGE_DIM, lib/file-attachments.ts:33; iOS MAX_IMAGE_DIM,
     *  Attachments.swift). Anthropic's vision pipeline downscales past ~1568px
     *  anyway, so this is the sweet spot: larger just inflates the payload without
     *  buying the model detail. Android sat at 1280px, so a chart/receipt/screenshot
     *  reached the agent SOFTER + smaller from Android than from web — the natives
     *  had drifted to match each other, not the reference (iOS realigned cycle 68). */
    private const val MAX_DIM = 1568
    /** JPEG quality for the model-bound image (web JPEG_QUALITY 0.85,
     *  lib/file-attachments.ts:34; iOS MODEL_IMAGE_QUALITY 0.85). Android sat at 70
     *  — visibly softer than web's 85 on the same photo; screenshot text lost
     *  crispness the model then had to guess at. */
    private const val QUALITY = 85

    // Server DOC_EXT_TO_FORMAT (mime path first, then extension fallback).
    private val EXT_TO_FORMAT = mapOf(
        "pdf" to "pdf", "csv" to "csv", "doc" to "doc", "docx" to "docx",
        "xls" to "xls", "xlsx" to "xlsx", "html" to "html", "htm" to "html",
        "txt" to "txt", "md" to "md", "markdown" to "md", "json" to "json", "xml" to "xml",
    )
    private val MIME_TO_FORMAT = mapOf(
        "application/pdf" to "pdf",
        "text/csv" to "csv",
        "application/msword" to "doc",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document" to "docx",
        "application/vnd.ms-excel" to "xls",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" to "xlsx",
        "text/html" to "html",
        "text/xml" to "xml", "application/xml" to "xml",
        "text/plain" to "txt", "text/markdown" to "md", "application/json" to "json",
    )

    /** SAF MIME filter for the document picker (superset of the format map). */
    val DOC_MIME_TYPES = arrayOf(
        "application/pdf", "text/csv", "text/plain", "text/markdown", "text/html",
        "application/json", "text/xml", "application/xml",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/vnd.ms-excel",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    )

    fun encode(context: Context, uri: Uri): String? = runCatching {
        // Buffer the source bytes once: we decode twice (bounds, then pixels) and
        // read EXIF, and a content:// stream isn't guaranteed re-openable/seekable.
        // A camera JPEG on disk is a few MB — trivial next to the decoded bitmap.
        val raw = context.contentResolver.openInputStream(uri)?.use { it.readBytes() } ?: return null

        // Pass 1: bounds only (no pixel allocation) → pick inSampleSize so the DECODED
        // bitmap lands near MAX_DIM. Decoding a Pixel 50MP shot full-res first would
        // allocate ~200MB of ARGB_8888 just to downscale it — an OOM waiting to happen
        // (iOS's UIImage pipeline is lazy about this; BitmapFactory is not). inSampleSize
        // downsamples AT decode time, so peak memory is bounded to the sampled size.
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(raw, 0, raw.size, bounds)
        val opts = BitmapFactory.Options().apply {
            inSampleSize = sampleSizeFor(bounds.outWidth, bounds.outHeight, MAX_DIM)
        }
        val decoded = BitmapFactory.decodeByteArray(raw, 0, raw.size, opts) ?: return null

        // Fine scale to the exact long-edge cap (inSampleSize only halves), then bake in
        // EXIF orientation — BitmapFactory ignores the flag, so a portrait phone photo
        // would otherwise reach the model rotated (iOS jpegData applies orientation for us).
        val scale = minOf(1f, MAX_DIM.toFloat() / maxOf(decoded.width, decoded.height, 1))
        val orientation = runCatching {
            ExifInterface(ByteArrayInputStream(raw)).getAttributeInt(
                ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL,
            )
        }.getOrDefault(ExifInterface.ORIENTATION_NORMAL)
        val matrix = orientationMatrix(orientation).apply { if (scale < 1f) preScale(scale, scale) }
        val bitmap = if (matrix.isIdentity) decoded
            else Bitmap.createBitmap(decoded, 0, 0, decoded.width, decoded.height, matrix, true)

        val out = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, QUALITY, out)
        Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)
    }.getOrNull()

    /** Long-edge cap + JPEG quality for the persisted bubble/history thumbnail (iOS 96px q0.6). */
    private const val THUMB_DIM = 96
    private const val THUMB_QUALITY = 60

    /**
     * A tiny 96px JPEG thumbnail (base64) from an already-encoded full-res image
     * base64 — the durable preview the user bubble + history render (iOS PendingAttachment
     * .thumb). Full payloads never persist (they'd bloat chat-history-<tiny>.json); only
     * these ~few-KB thumbs do. Returns null if the input can't be decoded (the bubble then
     * just shows text — never a broken tile), so it's safe on any stored string.
     */
    fun thumbnail(fullBase64: String): String? = runCatching {
        val bytes = Base64.decode(fullBase64, Base64.DEFAULT)
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)
        val opts = BitmapFactory.Options().apply {
            inSampleSize = sampleSizeFor(bounds.outWidth, bounds.outHeight, THUMB_DIM)
        }
        val decoded = BitmapFactory.decodeByteArray(bytes, 0, bytes.size, opts) ?: return null
        // The full-res base64 already baked in EXIF orientation (see encode), so the thumb
        // just needs the fine downscale to the exact long-edge cap.
        val scale = minOf(1f, THUMB_DIM.toFloat() / maxOf(decoded.width, decoded.height, 1))
        val bitmap = if (scale < 1f) {
            val m = Matrix().apply { preScale(scale, scale) }
            Bitmap.createBitmap(decoded, 0, 0, decoded.width, decoded.height, m, true)
        } else decoded
        val out = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, THUMB_QUALITY, out)
        Base64.encodeToString(out.toByteArray(), Base64.NO_WRAP)
    }.getOrNull()

    /** Largest power-of-two sample size keeping the sampled long edge ≥ [target]. */
    private fun sampleSizeFor(w: Int, h: Int, target: Int): Int {
        if (w <= 0 || h <= 0) return 1
        var sample = 1
        var longEdge = maxOf(w, h)
        while (longEdge / 2 >= target) { longEdge /= 2; sample *= 2 }
        return sample
    }

    /** EXIF orientation → transform matrix (rotation + mirror); identity for NORMAL. */
    private fun orientationMatrix(orientation: Int): Matrix = Matrix().apply {
        when (orientation) {
            ExifInterface.ORIENTATION_ROTATE_90 -> postRotate(90f)
            ExifInterface.ORIENTATION_ROTATE_180 -> postRotate(180f)
            ExifInterface.ORIENTATION_ROTATE_270 -> postRotate(270f)
            ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> postScale(-1f, 1f)
            ExifInterface.ORIENTATION_FLIP_VERTICAL -> postScale(1f, -1f)
            ExifInterface.ORIENTATION_TRANSPOSE -> { postRotate(90f); postScale(-1f, 1f) }
            ExifInterface.ORIENTATION_TRANSVERSE -> { postRotate(270f); postScale(-1f, 1f) }
        }
    }

    sealed class DocResult {
        data class Ok(val doc: PendingDoc) : DocResult()
        data class Err(val message: String) : DocResult()
    }

    /** Read a picked document, resolve its format, size-check, base64-encode. */
    fun encodeDocument(context: Context, uri: Uri): DocResult {
        val (displayName, size) = queryNameAndSize(context, uri)
        val mime = context.contentResolver.getType(uri)
        val format = formatFor(mime, displayName)
            ?: return DocResult.Err("${displayName ?: "file"} — unsupported document type")
        if (size != null && size > MAX_DOC_BYTES) {
            return DocResult.Err(
                "${displayName ?: "file"} is ${String.format(java.util.Locale.US, "%.1f", size / 1_048_576.0)}MB — documents must be under $MAX_DOC_LABEL",
            )
        }
        val bytes = runCatching {
            context.contentResolver.openInputStream(uri)?.use { it.readBytes() }
        }.getOrNull() ?: return DocResult.Err("Couldn't read ${displayName ?: "file"} — unsupported or corrupted")
        if (bytes.size > MAX_DOC_BYTES) {
            return DocResult.Err(
                "${displayName ?: "file"} is ${String.format(java.util.Locale.US, "%.1f", bytes.size / 1_048_576.0)}MB — documents must be under $MAX_DOC_LABEL",
            )
        }
        val b64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
        return DocResult.Ok(PendingDoc(sanitizeName(displayName), format, b64))
    }

    /** Server getDocumentFormat: MIME map first, extension fallback. internal so the
     *  resolution order (a mislabeled .pdf served as octet-stream still resolves by
     *  extension) is JVM-unit-testable. */
    internal fun formatFor(mime: String?, name: String?): String? {
        MIME_TO_FORMAT[mime]?.let { return it }
        val ext = name?.substringAfterLast('.', "")?.lowercase().orEmpty()
        return EXT_TO_FORMAT[ext]
    }

    /** Server buildContentBlocks name rule: strip ext, non-safe→_, cap 200, fallback.
     *  Uses the SAME regex as web (`\.[^.]+$`) — the prior substringBeforeLast diverged
     *  on a trailing-dot name ("archive." → web "archive_" but old Android "archive"),
     *  and this string is sent to Anthropic verbatim so it must match byte-for-byte.
     *  internal for parity tests. */
    internal fun sanitizeName(raw: String?): String {
        val base = (raw ?: "document").replace(Regex("""\.[^.]+$"""), "")
        val cleaned = base.replace(Regex("[^a-zA-Z0-9\\s\\-()\\[\\]]"), "_").take(200)
        return cleaned.ifBlank { "document" }
    }

    private fun queryNameAndSize(context: Context, uri: Uri): Pair<String?, Long?> {
        var name: String? = null
        var size: Long? = null
        runCatching {
            context.contentResolver.query(uri, null, null, null, null)?.use { c ->
                if (c.moveToFirst()) {
                    val ni = c.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                    if (ni >= 0 && !c.isNull(ni)) name = c.getString(ni)
                    val si = c.getColumnIndex(OpenableColumns.SIZE)
                    if (si >= 0 && !c.isNull(si)) size = c.getLong(si)
                }
            }
        }
        return name to size
    }

    /**
     * The multi-select cap decision: how many of a picker's [pickedCount] items fit
     * given [pendingCount] already staged, and whether any overflowed. The photo picker's
     * maxItems is the TOTAL cap (blind to what's already pending), so a pick can overshoot
     * the free slots — [ClampResult.accept] is the count to keep and [ClampResult.overflowed]
     * drives the "up to MAX attachments" message. Pure so the arithmetic (free = MAX −
     * pending, take(free)) is unit-tested once instead of living only inline in the picker
     * callback. iOS parity: AttachmentCodec.routeDrop's `capacity` walk + PhotosPicker
     * maxSelectionCount = free slots. [max] injectable for tests; defaults to [MAX].
     */
    data class ClampResult(val accept: Int, val overflowed: Boolean) {
        /** No slots left at all — the caller shows the cap message and ingests nothing. */
        val full: Boolean get() = accept == 0
    }

    fun clampPicks(pendingCount: Int, pickedCount: Int, max: Int = MAX): ClampResult {
        val free = (max - pendingCount).coerceAtLeast(0)
        val accept = pickedCount.coerceIn(0, free)
        // Overflow only counts when something was actually dropped — an empty pick,
        // or a pick that fits exactly, never reads as "you hit the cap".
        return ClampResult(accept = accept, overflowed = pickedCount > free)
    }

    /**
     * Decoded byte size of a base64 model payload (server `base64Bytes`, iOS
     * `payloadBytes`): base64 inflates 4/3×, so the raw bytes are length × 0.75.
     * The 96px history thumb is never sent, so it doesn't count here.
     */
    fun payloadBytes(base64: String): Int = (base64.length * 0.75).toInt()

    /** Sum of the model-bound payloads for a staged image/doc set (server `attachmentsPayloadBytes`). */
    fun payloadBytesOf(imagesBase64: List<String>, docs: List<PendingDoc>): Int =
        imagesBase64.sumOf { payloadBytes(it) } + docs.sumOf { payloadBytes(it.base64) }

    /**
     * Whether adding `addBase64` to the already-staged set stays within
     * MAX_PAYLOAD_BYTES — the batch total guard mirroring iOS `appendAttachment`.
     * Pure so the composer can gate an add up front with a "remove some first"
     * message instead of letting the worker reject the whole send.
     */
    fun fitsPayload(
        imagesBase64: List<String>,
        docs: List<PendingDoc>,
        addBase64: String,
        max: Int = MAX_PAYLOAD_BYTES,
    ): Boolean = payloadBytesOf(imagesBase64, docs) + payloadBytes(addBase64) <= max

    fun blocks(text: String, imagesBase64: List<String>, docs: List<PendingDoc> = emptyList()): JSONArray {
        val arr = JSONArray()
        arr.put(JSONObject().put("text", text.ifBlank { "Have a look." }))
        imagesBase64.take(MAX).forEach { b64 ->
            arr.put(
                JSONObject().put(
                    "image",
                    JSONObject()
                        .put("format", "jpeg")
                        .put("source", JSONObject().put("bytes", b64)),
                )
            )
        }
        docs.forEach { doc ->
            arr.put(
                JSONObject().put(
                    "document",
                    JSONObject()
                        .put("name", doc.name)
                        .put("format", doc.format)
                        .put("source", JSONObject().put("bytes", doc.base64)),
                )
            )
        }
        return arr
    }
}

File: ios/Tiny/Sources/Attachments.swift
/**
 * Attachments — web parity for photos-to-the-agent, on the device that
 * actually has the camera. Library picks ride PhotosPicker; camera rides
 * UIImagePickerController. Images downscale to ≤1568px JPEG (q0.85, web parity)
 * and travel as the same Converse content blocks the web sends:
 * { image: { format: "jpeg", source: { bytes: <base64> } } }.
 *
 * History persists only tiny 96px thumbs (web pattern: base64 payloads are
 * stripped on persist — a photo would blow the store).
 */
import SwiftUI
import PhotosUI

struct PendingAttachment: Identifiable, Equatable {
    let id = UUID()
    /// ≤1568px JPEG or raw document bytes, base64 — what the agent sees
    let base64: String
    /// 96px JPEG, base64 — bubble/history preview (images only)
    let thumb: String?
    /// Documents only: display name (extension stripped) + Converse format
    let docName: String?
    let docFormat: String?

    var isImage: Bool { docFormat == nil }

    /// Decoded byte size of the model-bound payload (base64 inflates 4/3×, so
    /// the raw bytes are length × 0.75) — web `base64Bytes` (lib/file-
    /// attachments.ts). The 96px thumb is history-only and never sent, so it
    /// doesn't count toward the request budget.
    var payloadBytes: Int { Int(Double(base64.count) * 0.75) }

    init(base64: String, thumb: String? = nil, docName: String? = nil, docFormat: String? = nil) {
        self.base64 = base64
        self.thumb = thumb
        self.docName = docName
        self.docFormat = docFormat
    }
}

/// Cap per message — Converse payloads have limits and mobile data is real
let MAX_ATTACHMENTS = 4

/// Per-document byte cap — the canonical cross-client limit (web
/// MAX_DOCUMENT_BYTES, lib/file-attachments.ts:38; Android MAX_DOC_BYTES,
/// Attachments.kt:28). iOS previously hardcoded 2.5MB here and rejected
/// 2.5–3.0MB docs the other clients accept.
let MAX_DOCUMENT_BYTES = 3_000_000

/// Total decoded-payload cap across all staged attachments (web
/// MAX_PAYLOAD_BYTES, lib/file-attachments.ts): base64 inflates 4/3× and the
/// history + system context ride along in the same request, so the body must
/// stay comfortably under the worker's ~4.5MB cap. The per-item caps
/// (3MB/doc, downscaled images) don't stop FOUR large picks from summing past
/// it — this is the batch guard the mobile clients were missing (only web had
/// it), so a too-heavy set was caught server-side as a send failure instead of
/// up front in the composer.
let MAX_ATTACHMENTS_PAYLOAD_BYTES = 3_500_000

extension Array where Element == PendingAttachment {
    /// Sum of the model-bound payloads (web `attachmentsPayloadBytes`).
    var payloadBytes: Int { reduce(0) { $0 + $1.payloadBytes } }
}

/// Longest edge for the model-bound image — the canonical cross-client value
/// (web MAX_IMAGE_DIM, lib/file-attachments.ts:33). Anthropic's vision pipeline
/// downscales past ~1568px anyway, so this is the sweet spot: any larger just
/// inflates the payload without buying the model detail. iOS + Android both sat
/// at 1280px, so a chart/receipt/screenshot reached the agent SOFTER and smaller
/// from mobile than from web — the natives had drifted to match each other, not
/// the reference.
let MAX_IMAGE_DIM: CGFloat = 1568
/// JPEG quality for the model-bound image (web JPEG_QUALITY, lib/file-
/// attachments.ts:34). iOS sat at 0.7 — visibly softer than web's 0.85 on the
/// same photo; text in a screenshot lost crispness the model then had to guess at.
let MODEL_IMAGE_QUALITY: CGFloat = 0.85

enum AttachmentCodec {
    static func encode(_ image: UIImage) -> PendingAttachment? {
        // Model-bound image at web parity (1568px / q0.85). The 96px thumb is a
        // history-only preview — deliberately tiny so the persisted store stays
        // light (never sent to the model), so it keeps its own low settings.
        guard let full = downscale(image, maxDim: MAX_IMAGE_DIM).jpegData(compressionQuality: MODEL_IMAGE_QUALITY),
              let thumb = downscale(image, maxDim: 96).jpegData(compressionQuality: 0.6)
        else { return nil }
        return PendingAttachment(base64: full.base64EncodedString(), thumb: thumb.base64EncodedString())
    }

    static func downscale(_ image: UIImage, maxDim: CGFloat) -> UIImage {
        let size = image.size
        let scale = min(1, maxDim / max(size.width, size.height, 1))
        guard scale < 1 else { return image }
        let target = CGSize(width: size.width * scale, height: size.height * scale)
        return UIGraphicsImageRenderer(size: target).image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
    }

    /// Outcome of a document pick — mirrors Android's `DocResult` (Attachments.kt)
    /// so a rejected doc surfaces WHY in the composer banner instead of vanishing.
    /// iOS previously returned `PendingAttachment?`; a `nil` from an oversize or
    /// unreadable pick was silently swallowed by the file-importer/drop callers,
    /// while web toasts a named error and Android sets `vm.error`. Now the reject
    /// reason rides `.err(message)`, phrased to match web (lib/file-attachments.ts)
    /// and Android word-for-word ("<file> is X.XMB — documents must be under 2.9MB").
    enum DocResult: Equatable {
        case ok(PendingAttachment)
        case err(String)
    }

    /// Byte cap rendered for reject copy — "2.9MB": web renders the cap in MiB
    /// (`(MAX_DOCUMENT_BYTES/1024/1024).toFixed(1)` → "2.9MB" for 3_000_000) and
    /// Android's MAX_DOC_LABEL computes the same. Derived, not hardcoded, so the
    /// copy self-updates (and stays consistent with the file-size figure in the
    /// same sentence, which is also MiB) if MAX_DOCUMENT_BYTES moves.
    private static var docCapLabel: String {
        String(format: "%.1fMB", Double(MAX_DOCUMENT_BYTES) / 1_048_576)
    }

    /// Documents (PDF/CSV/…): raw bytes → Converse document block fields.
    /// Same shape the web sends: { document: { name, format, source: { bytes } } }
    static func encodeDocument(url: URL) -> DocResult {
        // File URLs only — iPad drag-and-drop can hand over https URLs
        // (Safari link drags); Data(contentsOf:) would synchronously fetch
        // those over the network on the main thread. Web content isn't a
        // document attachment; reject the scheme, not just the size. (The drop
        // router already sends links to the composer, so this is defensive.)
        guard url.isFileURL else { return .err("Only files can be attached") }
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }
        let name = url.lastPathComponent  // full filename incl. extension (web `file.name`)
        // Couldn't-read / empty — web: "Couldn't read <name> — unsupported or
        // corrupted"; Android: "couldn't read <file>".
        guard let data = try? Data(contentsOf: url), !data.isEmpty else {
            return .err("Couldn't read \(name) — unsupported or corrupted")
        }
        // 3MB per-document cap (MAX_DOCUMENT_BYTES) — the canonical cross-client
        // limit. Was a silent nil; now names the file + its size like web/Android.
        guard data.count <= MAX_DOCUMENT_BYTES else {
            let mb = Double(data.count) / 1_048_576
            return .err(String(format: "%@ is %.1fMB — documents must be under %@", name, mb, docCapLabel))
        }
        // Extension → Converse doc format. Mirrors web's accept list + Android's
        // EXT_TO_FORMAT (Attachments.kt): htm folds to html, markdown to md, and
        // json/xml are first-class — before this they fell through to "txt", so a
        // picked .json reached the agent mislabeled as plain text.
        let formats = ["pdf": "pdf", "csv": "csv", "doc": "doc", "docx": "docx",
                       "xls": "xls", "xlsx": "xlsx", "html": "html", "htm": "html",
                       "md": "md", "markdown": "md", "txt": "txt", "json": "json", "xml": "xml"]
        let format = formats[url.pathExtension.lowercased()] ?? "txt"
        let docName = url.deletingPathExtension().lastPathComponent
        return .ok(PendingAttachment(base64: data.base64EncodedString(), docName: docName, docFormat: format))
    }

    /// Converse content blocks: text first (never empty — web sends
    /// "Have a look." for image-only messages), then one block per attachment
    static func blocks(text: String, attachments: [PendingAttachment]) -> [[String: Any]] {
        var out: [[String: Any]] = [["text": text.isEmpty ? "Have a look." : text]]
        for att in attachments.prefix(MAX_ATTACHMENTS) {
            if let format = att.docFormat {
                // Name rules mirror the web: safe charset, ≤200 chars
                let safe = (att.docName ?? "document")
                    .replacingOccurrences(of: "[^a-zA-Z0-9\\s\\-()\\[\\]]", with: "_", options: .regularExpression)
                    .prefix(200)
                out.append(["document": ["name": safe.isEmpty ? "document" : String(safe), "format": format, "source": ["bytes": att.base64]]])
            } else {
                out.append(["image": ["format": "jpeg", "source": ["bytes": att.base64]]])
            }
        }
        return out
    }
}

/// What a dropped URL should become — pure decision, no I/O, testable.
/// (Views.swift's dropDestination executes these; TinyTests covers the
/// routing table that was previously inline-in-closure and untestable.)
enum DropIntake: Equatable {
    /// file:// URL with attachment capacity left → encodeDocument path
    case document(URL)
    /// https/mailto/… link drag → append to the composer draft
    case composerText(String)
    /// file:// URL but attachments are full → dropped on the floor
    case overCapacity
}

extension AttachmentCodec {
    /// Route dropped URLs: files become documents while capacity lasts
    /// (MAX_ATTACHMENTS cap counts existing pending + accepted-this-drop),
    /// non-file links become composer text regardless of capacity.
    static func routeDrop(urls: [URL], pendingCount: Int, max: Int = MAX_ATTACHMENTS) -> [DropIntake] {
        var capacity = max - pendingCount
        return urls.map { url in
            if url.isFileURL {
                guard capacity > 0 else { return .overCapacity }
                capacity -= 1
                return .document(url)
            }
            return .composerText(url.absoluteString)
        }
    }

    /// Merge a dragged link into the composer draft (single space seam,
    /// no leading space on an empty draft).
    static func mergeLink(_ draft: String, _ link: String) -> String {
        draft.isEmpty ? link : draft + " " + link
    }
}

/// Doc chip for the pending strip + history bubbles
struct DocChip: View {
    let name: String
    // The tiny's accent is the ONLY brand color (Theme.swift) — web tints the
    // attachment chips with rgba(var(--tiny-accent-rgb),…), so a hardcoded green
    // clashed on every non-green tiny. Read it from the environment ChatView
    // already injects (Views.swift .environment(\.tinyAccent, chat.accent)).
    @Environment(\.tinyAccent) private var accent

    var body: some View {
        Label(name, systemImage: "doc.fill")
            .font(.caption)
            .lineLimit(1)
            .padding(.horizontal, 10).padding(.vertical, 6)
            .background(accent.opacity(0.12), in: Capsule())
            .foregroundStyle(accent)
    }
}

/// UIImagePickerController wrapper — PhotosPicker has no camera source
struct CameraPicker: UIViewControllerRepresentable {
    let onImage: (UIImage) -> Void
    @Environment(\.dismiss) private var dismiss

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(self) }

    final class Coordinator: NSObject, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        let parent: CameraPicker
        init(_ parent: CameraPicker) { self.parent = parent }

        func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
            if let image = info[.originalImage] as? UIImage { parent.onImage(image) }
            parent.dismiss()
        }

        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            parent.dismiss()
        }
    }
}

/// Thumbnail from a stored base64 thumb (bubbles + pending strip)
struct AttachmentThumb: View {
    let base64: String
    var size: CGFloat = 56
    // Accent-tinted border to match web (rgba(var(--tiny-accent-rgb),0.3)) and
    // the accent-themed composer/chips — not a hardcoded green on every tiny.
    @Environment(\.tinyAccent) private var accent

    var body: some View {
        if let data = Data(base64Encoded: base64), let img = UIImage(data: data) {
            Image(uiImage: img)
                .resizable()
                .aspectRatio(contentMode: .fill)
                .frame(width: size, height: size)
                .clipShape(RoundedRectangle(cornerRadius: 10))
                .overlay(RoundedRectangle(cornerRadius: 10).stroke(accent.opacity(0.3), lineWidth: 1))
        }
    }
}

// ── Share-link plumbing (conversation share links) ─────────────────────────

extension URL: @retroactive Identifiable {
    public var id: String { absoluteString }
}

/// UIActivityViewController wrapped for .sheet — ShareLink can't present
/// from a Menu action's async completion, this can.
struct ActivitySheet: UIViewControllerRepresentable {
    let items: [Any]
    func makeUIViewController(context: Context) -> UIActivityViewController {
        UIActivityViewController(activityItems: items, applicationActivities: nil)
    }
    func updateUIViewController(_ vc: UIActivityViewController, context: Context) {}
}
